src/parkapi_sources/converters/apcoa/validators.py

```diff
@@ -0,0 +1,125 @@
+
+
+class ApcoaCarparkType(Enum):
+    MLCP = 'MLCP'
+    OFF_STREET_OPEN = 'Off-street open'
+    OFF_STREET_UNDERGROUND = 'Off-street underground'
+    ON_STREET = 'On-street'
+    OPEN_SURFACE = 'Open Surface'
+
+    def to_parking_site_type_input(self) -> ParkingSiteType:
+        # TODO: find out more details about this enumeration for a proper mapping
+        return {
+            self.MLCP: ParkingSiteType.OFF_STREET_PARKING_GROUND,
+            self.OFF_STREET_OPEN: ParkingSiteType.OFF_STREET_PARKING_GROUND,
+            self.OFF_STREET_UNDERGROUND: ParkingSiteType.UNDERGROUND,
+            self.ON_STREET: ParkingSiteType.ON_STREET,
+            self.OPEN_SURFACE: ParkingSiteType.OFF_STREET_PARKING_GROUND,
+        }.get(self, ParkingSiteType.OTHER)
+
+
+class ApcoaNavigationLocationType:
+    DEFAULT = 'default'
+    CAR_ENTRY = 'CarEntry'
+
+
+class ApcoaOpeningHoursWeekdays:
+    Weekdays = ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday']
+    OpeningTime = '00:00 - 00:00'
+
+
+@validataclass
+class ApcoaCarparkTypeNameInput:
+    Name: ApcoaCarparkType = EnumValidator(ApcoaCarparkType)
+
+
+@validataclass
+class ApcoaLocationGeocoordinatesInput:
+    Longitude: Decimal = NumericValidator()
+    Latitude: Decimal = NumericValidator()
+
+
+@validataclass
+class ApcoaNavigationLocationsInput:
+    GeoCoordinates: ApcoaLocationGeocoordinatesInput = DataclassValidator(ApcoaLocationGeocoordinatesInput)
+    LocationType: str = StringValidator()
+
+
+@validataclass
+class ApcoaAdressInput:
+    Street: Optional[str] = Noneable(StringValidator())
+    Zip: Optional[str] = Noneable(StringValidator())
+    City: Optional[str] = Noneable(StringValidator())
+    Region: Optional[str] = Noneable(StringValidator())
+
+
+@validataclass
+class ApcoaParkingSpaceInput:
+    Type: str = StringValidator()
+    Count: int = IntegerValidator(allow_strings=True, min_value=0)
+
+
+@validataclass
+class ApcoaOpeningHoursWeekdayInput:
+    Weekday: str = StringValidator()
+    OpeningTimes: str = StringValidator()
+
+
+@validataclass
+class ApcoaParkingSiteInput:
+    CarParkId: int = IntegerValidator(allow_strings=True)
+    CarparkLongName: Optional[str] = Noneable(StringValidator())
+    CarparkShortName: Optional[str] = Noneable(StringValidator())
+    CarParkWebsiteURL: Optional[str] = Noneable(UrlValidator())
+    CarparkType: ApcoaCarparkTypeNameInput = DataclassValidator(ApcoaCarparkTypeNameInput)
+    Address: ApcoaAdressInput = DataclassValidator(ApcoaAdressInput)
+    NavigationLocations: list[ApcoaNavigationLocationsInput] = ListValidator(DataclassValidator(ApcoaNavigationLocationsInput))
+    Spaces: list[ApcoaParkingSpaceInput] = ListValidator(DataclassValidator(ApcoaParkingSpaceInput))
+    OpeningHours: list[ApcoaOpeningHoursWeekdayInput] = ListValidator(DataclassValidator(ApcoaOpeningHoursWeekdayInput))
+
+    # TODO: ignored multiple attributes which do not matter so far
```

Comment:
As we support `photo_url`, you could add `CarParkPhotoURLs` to your validation and mapping.

Reply:
If a value is null, you can use `Noneable` for accepting `None`: `foo: Optional[str] = Noneable(StringValidator())`

Reply:
Ok amended